### DISCUSSION TITLE
Setting up LEDStrip

### DISCUSSION THREAD

**Discussion by MysDevLop** [Pi4J/pi4j#318] [Q&A]
I see that there are four .so files in your library. I have added these to the classpath, but I do not explicitly load them. I get no console message about them.

I use this code to setup an LEDStrip
```
 DigitalOutputConfigBuilder ledConfig;
 Context         ivPI4J;
    LedStrip        ivLEDStrip;

ivPI4J    = Pi4J.newAutoContext();

 ledConfig = DigitalOutput.newConfigBuilder( ivPI4J );

  ledConfig.id( "ID: " + ivPrinter.getUserName() );
   ledConfig.name( ivPrinter.getUserName() );
   ledConfig.address( ivLEDOutput.getLEDOutputParm().getPinNumber().getPinOut().getPMWChannel() );
        ledConfig.shutdown( DigitalState.LOW );

        ledConfig.build();

        ivPI4J.create( ledConfig );

        ivLEDStrip = new LedStrip( ivPI4J, ivPrinter.getUserName(), colors.length, ivLEDOutput.getLEDOutputParm().getPinNumber().getPinOut().getPMWChannel().intValue(), ivPrinter.getLEDType() );

```
When I start the program I get this in the console:

```
[main] INFO com.pi4j.Pi4J - New auto context
[main] INFO com.pi4j.Pi4J - New context builder
[main] INFO com.pi4j.platform.impl.DefaultRuntimePlatforms - adding platform to managed platform map [id=raspberrypi; name=RaspberryPi Platform; priority=5; class=com.pi4j.plugin.raspberrypi.platform.RaspberryPiPlatform]

```

Is this correct?

I am not getting ANY output to my LED strip. I have tried both PMW pins.

**Comment by MysDevLop** [comment]
I should add that I modified the LedStrinp constructor to accept the chip type:

```
 public LedStrip( Context pi4j, String printerName, int numberOfLEDs, int channel, LEDChipType chipType )
    {
        super( pi4j, Spi.newConfigBuilder( pi4j ).id( "SPI-" + channel ).bus( SpiBus.BUS_1 ).name( "LED Strip: " + printerName ).address( Integer.valueOf( channel ) ).baud( chipType.getBaud() ).build() );

```

**Comment by MysDevLop** [comment]
Ok, I fiddled with stuff, and now I am getting the error during ```.so``` file load:

```
can't load this .so (machine code =0x3f) on a AARCH64 platform
```

This seems to read that the ```.so``` file is compiled for a different processor?

I am using a Raspberry PI 4, model B with 2GB of ram.

**Comment by MysDevLop** [comment]
Ok, I have better error messages (copy paste from remote terminal). 

```
String libRoot =  System.getProperty( "user.dir" ) + "/libs/";

System.load( libRoot + "pi4j-library-linuxfs-aarch64.so" );
System.load( libRoot + "pi4j-library-linuxfs-armhf.so" );
System.load( libRoot + "pi4j-library-pigpio-aarch64.so" );
System.load( libRoot + "pi4j-library-pigpio-armhf.so" );
 ```
Error messages:
```
libs/pi4j-library-linuxfs-aarch64.so: libs/pi4j-library-linuxfs-aarch64.so: cannot open shared object file: No such file or directory (Possible cause: can't load this .so (machine code=0x3f) on a AARCH64 platform)

libs/pi4j-library-linuxfs-armhf.so: libs/pi4j-library-linuxfs-armhf.so: wrong ELF class: ELFCLASS32 (Possible cause: can't load ARM .so on a AARCH64 platform)

libs/pi4j-library-pigpio-aarch64.so: libs/pi4j-library-pigpio-aarch64.so: cannot open shared object file: No such file or directory (Possible cause: can't load this .so (machine code=0x3f) on a AARCH64 platform)

libs/pi4j-library-pigpio-armhf.so:libs/pi4j-library-pigpio-armhf.so: wrong ELF class: ELFCLASS32 (Possible cause: can't load ARM .so on a AARCH64 platform)
```

Any help would be much appreciated

**Comment by MysDevLop** [comment]
Could it be because I am not using JBANG?

I have a Java Swing application in a .jar file which I start with ```sudo java -jar app.jar```

It runs, and when I have debug turned on I get ```FINE: Finished rendering of LED strip [time]``` messages.

As far as I can tell, I am using identical code to yours. The in-data on the LED string is connected to ```GPIO 12 (PWM0)``` which should be ```channel 0``` in your code.

I can see the ```/lib/aarch64/libpi4j-pigpio.so``` in the ```C:\Eclipse\addon\pi4\pi4j-library-pigpio.jar```. I have to assume it is being loaded.

I really do not understand why this does not work.

**Comment by MysDevLop** [comment]
Here is the totality of the console messages:
```
[main] INFO com.pi4j.Pi4J - New auto context
[main] INFO com.pi4j.Pi4J - New context builder
[main] INFO com.pi4j.platform.impl.DefaultRuntimePlatforms - adding platform to managed platform map [id=raspberrypi; name=RaspberryPi Platform; priority=5; class=com.pi4j.plugin.raspberrypi.platform.RaspberryPiPlatform]
FINE: SPI is open [12:59:57 PM]
FINE: Finished rendering of LED strip [12:59:57 PM]
FINE: Finished rendering of LED strip [12:59:57 PM]
FINE: Finished rendering of LED strip [12:59:57 PM]
FINE: Finished rendering of LED strip [12:59:57 PM]
FINE: LED strip with 50 LEDs [12:59:58 PM]
FINE: Finished rendering of LED strip [1:00:00 PM]
FINE: Finished rendering of LED strip [1:00:02 PM]
FINE: Finished rendering of LED strip [1:00:04 PM]
FINE: Finished rendering of LED strip [1:00:06 PM]
FINE: Finished rendering of LED strip [1:00:08 PM]
FINE: Finished rendering of LED strip [1:00:10 PM]
FINE: Finished rendering of LED strip [1:00:12 PM]
```

The first four rendering message are from the ORANGE flash. My LED string has 50 LEDs. The LED string is updated about every 2 seconds, so that are the other rendering messages.

What am I missing here?

**Comment by taartspi** [comment]
I will ask you help us reset on your environment.  You mention PWM and the 
program output refers to SPI being open so we need to better understand 
your work  There are three posts pertaining to this work and several 
different examples were mentioned.
Please answer:

1. Are you using an example referenced in the pi4j website or our 
previous response

2. if (1) is yes, provide the link

3. If (1) is no,where did you obtain the code or possibly wrote it
yourself.

4. Assuming (1) is yes, how did you download the code

5. Assuming (1) is yes how did you build the code

6. If you modified the code, did you first execute the code before any modifications

7. Describe your devices connections to the Pi4  GPIO pins.

8. How do you execute the program, please copy paste the command or
describe action is using and IDE

9.Describe if the LEDs make any illumination when running the 
program 

10. Pipe all the output into a file so we can view all the messages 
generated when running the program.  Attach the file to your comment.


Hopefully with the answers for 1,2,3 we can see who can best help
you with your error.

**Comment by MysDevLop** [comment]
My application reads information from a server, then maps the information into an LED strip. I wrote the application on my Windows machine using Eclipse as the IDE. I then exported a .jar file, copied it to the raspberry pi (running the Pi4J image). The PI user has admin privileges.

```sudo java -jar app.jar```

My application starts, reads the server, then displays a sample of the output to the screen. So I know that the information is being gathered.

It also attempts to echo that to the LED strip.

I got the .jar files from Github. The LedStrip code (and associated files) I got from Github (com.pi4j.catalog.components package). The .jar files are:
```
pi4j-core.jar
pi4j-core-javadoc.jar
pi4j-core-sources.jar
pi4j-library-linuxfs.jar
pi4j-library-pigio.jar
pi4j-plugin-pigio.jar
pi4j-plugin-raspberrypi.jar
```
These are in the Eclipse project build path. The ```com.pi4j.catalog.components package``` is part of the source code path. In debug mode I can single step through all of it, including the ```render()```, until I reach the native code classes. I can see the color value which each LED is supposed to show.

I reverted all the code back to your stock code, and am running it with no modifications, other than boosting the brightness to 1.0f:
```
    private static final float DEFAULT_BRIGHTNESS    = 1.0f;
```
 I start the LedStrip:
- void refreshPrinter() is called whenever there is a change to the color scheme, so with no changes only called once.
- synchronized void show() is called every time information is obtained from the server, about every two seconds
```
public class UpdateLEDStrip
{ 
private Context  ivPI4J = null;
private LedStrip  ivLEDStrip;

  public void refreshPrinter()
    {
        PrinterColour     colour;
        LEDColourSet[]    colors;

        colour = ivPrinter.getPrinterColour();
        colors = colour.getLEDSet();

        if (ivPI4J != null)
        {
            ivPI4J.shutdown( "SPI-" + channel );
        }
     
     ivPI4J     = Pi4J.newAutoContext();
     ivLEDStrip = new LedStrip( ivPI4J, colors.length );
  }

 public synchronized void show()
    {
      PrinterColour  colour;
        LEDColourSet[] colors;
  
        colour = ivPrinter.getPrinterColour();
        colors = colour.getLEDSet();  // updated elsewhere

        for (int i = 0; i < colors.length; i++)
            ivLEDStrip.setPixelColor( i, colors[i].getLEDColour() );

        ivLEDStrip.render( Duration.ZERO );
    }
}
```
colors[i].getLEDColour() returns
```
 LedStrip.LedColor.createColorRGB( ivColour.getRed(), ivColour.getGreen(), ivColour.getBlue() );
```

I have tested the LED strip on an arduino and it works. I have measured the voltage at the end of the strip and it is +5 volts when connected to the Raspberry PI.

The data is connected initially to physical pin 32 (GPIO 12), then I tried physical pin 19 (GPIO 10), even though the LedStrip.java shows
```
private static final int  DEFAULT_SPI_CHANNEL   = 0;
```
which I understand to be ```GPIO 12 PWM0```.

I have no other devices connected, though eventually I will be using a multiplexor (not used right now, and all code for the multiplexor is commented out).

I got the sample code from [https://pi4j.com/examples/components/ledstrip/](url)

The LEDs do not show ANY lights, including the ORANGE flash which is shown during LedStrip initialization.

Setting the logger to FINE, I copied all the console information as per the above comment.

**Comment by MysDevLop** [comment]
I should add that using
```
sudo java -jar app.jar > op.txt
```
that op.txt has no output (0 bytes). Only the console has messages when I run it in a terminal.

**Comment by MysDevLop** [comment]
I have tried two types of LED, ```SMS/WS2811/RGB``` and ```WS2812B```

**Comment by taartspi** [comment]
I cloned :
pi@raspi464alt:~/Pi4J_V2/pi4j-example-components $ git remote -v
origin	https://github.com/Pi4J/pi4j-example-components (fetch)
origin	https://github.com/Pi4J/pi4j-example-components (push)

I believe this is the same code base you are using.

Built it and used their launcher to execute example 12, led strip.

The LED strip did not illuminate.

I mentioned I did a simple test program to operate a LED strip. If you look at 
my readme 
https://github.com/Pi4J/pi4j-example-devices/blob/master/src/main/java/com/pi4j/devices/neopixel94v/README.md
I explain limiting the peripheral speed was required. 

So I modified the  /boot/config.txt, rebooted and the LED strip works correctly.



You listed the PINs you tested, this example uses the SPI, so MOSI pin.
You wiring should match this:
https://pi4j.com/examples/components/ledstrip/


Note:  backup this file before modification, incorrectthe edits can result in
the Pi failing to boot.
 
 
This change it for the Pi4, you reboot after completing the edit.

PI4 : Add the following to  /boot/config.txt
core_freq=250
core_freq_min=250                                                                                                                                                     I did not review the console information, if the above change does not resolve your problem I will then look at that

**Comment by MysDevLop** [comment]
Ok, I modified ```config.txt``` and rebooted. I still don't get any LED lights. I tried with both LED types I have. I have the data-in on the strips connected to the MOSI pin.

That is the code base I am using. I downloaded the ZIP file Github generates, expanded it, then copied the ```com.pi4j.catalog.applications``` and ```com.pi4j.catalog.components``` code trees into my project. I also had to add the ```org.slf4j.jar```, ```slf4-api.jar```, and ```slf4-simple.jar``` jar files for the logging. Plus the ```picocli``` code.

I added the other ```pi4j-``` jar files as listed above.

The only messages I get are the ones from the console (shown above). Piping to a file ```op.txt``` gives an empty file, though if I add ```System.out.println()``` calls, then the ```op.txt``` contains those messages. At any rate, the logger messages go to the console.

I really want to thank you for this help. I am mostly running blind here as I have never worked with a PI system before, or with ```native``` functionality. It's a steep learning curve for me.  The only other experience I have with LED strips is on an arduino using the FASTLed library, which does work for me.

**Comment by taartspi** [comment]
Make certain raspi-config shows the SPI interface enabled.  Read all of this if the interface was NOT enabled.
                                                                                                                                                                                                                                     
 Ok,  you mentioned setting the default intensity, I did not see that in my source code, I did not really have the 
same source as you.  I found what you have, downloaded and built it.  Looking in the source code it uses the SPI bus 1.
The documentation diagram shows 0. (we need to fix this).

I removed the changes to  /boot/config.txt and rebooted

Connected the ledStrip to GPIO20 SPI_1 mosi     I do not understand how SPI_1 operates
without updating the /boot/config.txt with dtoverlay=spi1-3cs   But it does.  see https://pinout.xyz/pinout/spi
Using 'raspi-gpio get'  only spi_0 is listed.   
                                                                                                                                                                                                                                          
 I do not know why it was moved to SPI_1, you could change back to SPI_0 and test. Move the led strip connection

The led patterns function as described.
I also changed the intensity to 1.0f and number leds to 50 to test how you are using the code.

Suggest  Remove the core_freq updates from the /boot/config.txt reboot.
Use GPIO20 to drive the led strip (I believe you tried this earlier)

Still fails.
Assuming you don't have a data scope to see what bits are flowing from GPIO20, read that 
section on pinouts and try setting that dtoverlay, boot and test.    The SPI_1 functioning by default is baffling.

Still fails, can you use the downloaded code as an application.  Run it 
com.pi4j.example/com.pi4j.catalog.Launcher -d     I am not certain of your
environment to know how you would accomplish this. But it removes question of other code
and packaging.

**Reply by taartspi** [reply]
And this morning using what i believe the latest code and what u have:     That  code uses channel0 and spi-1. Works correctly.  Changed to spi-0.  Rebuilt. Works Correctly
  I don't know the reason for the change between 0 and 1

**Comment by MysDevLop** [comment]
I have tried various values in combination for:
- SpiBus
- channel
- baud (from 100 to 500,000)

And I have tried different GPIO pins.

I cannot get this to work. Strangely, if I rub my finger over the data-in on the strip, I get the LEDs to light in random colours. 

I do not understand the relationship between the ```channel``` and ```bus```. Are they used in tandem? Only one, the other is ignored? How is the GPIO pin determined? How can I set the output to a specific pin?

Also, from the reading I have done, only the two PMW pins (GPIO12 and GPIO13) are supposedly capable of the rapid changes in value which an LED strip requires.

When I used ```raspi-gpio get``` I got up to GPIO53! Yes, that is fifty-three. The GPIO are organized into three banks. BANK0 is from GPIO0 to GPIO27. BANK1 is from GPIO28 to GPIO45, and BANK1 is from GPIO46 to GPIO53.

Anyway, I still cannot get this to work. Again, this is Raspberry PI 4, MOD B, 2BG of RAM. If that makes a difference?

I am writing this app to read from a server, then show LED results. I am hoping to distribute this for free for other people to use, so I need it as robust as possible, and explainable to the end user.

**Comment by MysDevLop** [comment]
If you have later code/jar files, where can I get them from? So we are using the same code base.

Thanks

**Comment by taartspi** [comment]
The code is from the location we discussed earlier  https://github.com/Pi4J/pi4j-example-components             This is the latest code.  The channel attribute is the ChipSelect (CE0 or CE1)  This isn't relevant in this program as the LED strip has no concept of being selected.     The bus, is whether SPI_bus0 or SPI_bus1.     Looking in the **classs public class LedStrip extends SpiDevice,** It creates the SPI device and sets it operating frequency for a Pi4.  You are not required to set any other frequency values.   Near the top of this file a comment explains which GPIO is connected to the LED strip depending upon which SPI_bus is used.    I still think it best you make certain the SPI interface is enabled, then use this example application to validate the LEDs operate as described in the LedStrip_App.
